reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+  if (controller._insideUnderlyingByteSource) {
+    controller._callPullOrPullIntoLaterIfNeeded = true;
+    return;
+  }
+
+  Promise.resolve().then(CallPullOrPullIntoRepeatedlyIfNeeded.bind(undefined, controller));
+}
+
+function CallPullOrPullIntoRepeatedlyIfNeeded(controller) {
+  const stream = controller._controlledReadableByteStream;
+
+  while (true) {
+    if (!controller._callPullOrPullIntoLaterIfNeeded) {
+      return;
+    }
+
+    if (controller._closeRequested) {
+      return;
+    }
+    if (stream._state !== 'readable') {
+      return;
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined) {
+      return;
+    }
+
+    if (IsReadableByteStreamReader(reader)) {
+      if (reader._readRequests.length === 0) {
+        return;
+      }
+
+      CallPull(controller);
+    } else {
+      assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+      if (reader._readIntoRequests.length === 0) {
+        return;
+      }
+
+      CallPullInto(controller);
+    }
+  }
+}
+
+function CancelReadableByteStream(stream, reason) {
+  if (stream._state === 'closed') {
+    return Promise.resolve(undefined);
+  }
+  if (stream._state === 'errored') {
+    return Promise.reject(stream._storedError);
+  }
+
+  CloseReadableByteStream(stream);
+
+  const controller = stream._controller;
+
+  controller._totalQueuedBytes = 0;
+  controller._queue = [];
+
+  const sourceCancelPromise = PromiseInvokeOrNoop(controller._underlyingByteSource, 'cancel', [reason]);
+  return sourceCancelPromise.then(() => undefined);
+}
+
+function CloseReadableByteStream(stream) {
+  assert(IsReadableByteStream(stream), 'stream must be ReadableByteStream');
+  assert(stream._state === 'readable', 'state must be readable');
+
+  stream._state = 'closed';
+
+  if (IsReadableByteStreamLocked(stream)) {
+    CloseReadableByteStreamReader(stream._reader);
+  }
+}
+
+function CloseReadableByteStreamReader(reader) {
+  reader._state = 'closed';
+
+  reader._closedPromise_resolve(undefined);
+  reader._closedPromise_resolve = undefined;
+  reader._closedPromise_reject = undefined;
+
+  if (IsReadableByteStreamReader(reader)) {
+    for (const req of reader._readRequests) {
+      req.resolve(CreateIterResultObject(undefined, true));
+    }
+
+    reader._readRequests = [];
+  } else {
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    if (reader._readIntoRequests.length > 0) {
+      return;
+    }
+  }
+
+  DetachReadableByteStreamReader(reader);
+}
+
+function DestroyReadableByteStreamController(controller) {
+  controller._pendingPullIntos = []
+  controller._queue = [];
+}
+
+function DetachReadableByteStreamReader(reader) {
+  reader._ownerReadableByteStream._reader = undefined;
+  reader._ownerReadableByteStream = undefined;
+}
+
+function EnqueueInReadableByteStreamController(controller, chunk) {
```

Comment:
Yeah, I also wondered why I didn't use it, but just remembered. The BYOB reader may consume only a part of a chunk in a queue. Since the existing queue-with-sizes machinery is designed to work chunk-wise, we need to shift, update (partially consume), and then unshift.

See `FillPullIntoDescriptorFromQueue()`.